lib/resolvers/PackageResolver.ts

```diff
@@ -14,44 +15,55 @@ export class PackageResolver {
    * Returns a promise which resolves with the retrieved directories.
    */
   static resolve(version: string, cacheRoot: string): Promise<MaterialToolsPackage> {
-    let localSourcePath = '';
+    return this._resolveVersion(version).then(targetVersion => {
+      // Run validators for the current version
+      this._validateVersion(targetVersion);
 
-    if (version === 'local') {
-      // Update the resolving version to the retrieved local version.
-      version = this._retrieveLocalVersion();
-    }
+      return this._isExisting(path.join(path.resolve(cacheRoot), targetVersion))
+        .then(this._resolveDirectories)
+        .then(directories => {
+          Logger.info('Using Angular Material version from cache.');
 
-    // Run validators for the current version
-    this._validateVersion(version);
-
-    return this._isExisting(path.join(path.resolve(cacheRoot), version))
-      .then(this._resolveDirectories)
-      .then(directories => {
-        Logger.info('Using Angular Material version from cache.');
-
-        return {
-          root: localSourcePath || path.join(directories.module, '..'),
-          module: directories.module,
-          source: directories.source,
-          version: version
-        };
-      })
-      .catch(directories => {
-        directories = this._resolveDirectories(directories);
-
-        return Promise.all([
-          VersionDownloader.getModuleVersion(version, directories.module),
-          VersionDownloader.getSourceVersion(version, directories.source)
-        ]).then(downloadPaths => {
           return {
-            root: localSourcePath || path.join(directories.module, '..'),
-            module: downloadPaths[0],
-            source: downloadPaths[1],
-            version: version
+            root: path.join(directories.module, '..'),
+            module: directories.module,
+            source: directories.source,
+            version: targetVersion
           };
+        })
+        .catch(directories => {
+          directories = this._resolveDirectories(directories);
+
+          return Promise.all([
+            VersionDownloader.getModuleVersion(targetVersion, directories.module),
+            VersionDownloader.getSourceVersion(targetVersion, directories.source)
+          ]).then(downloadPaths => {
+            return {
+              root: path.join(directories.module, '..'),
+              module: downloadPaths[0],
+              source: downloadPaths[1],
+              version: targetVersion
+            };
+          });
         });
+    });
+  }
 
-      });
+  /**
+   * Processes the different version types.
+   */
+  private static _resolveVersion(version: string): Promise<string> {
+    return new Promise((resolve, reject) => {
```

Comment:
This should be simpler

```ts
if (version === 'latest') {
  return this._retrieveLatestVersion();
} else if (version === 'local') {
  return Promise.resolve(this._retrieveLocalVersion());
} else {
  // FYI: `Promise.resolve` is pretty efficient.
  return Promise.resolve(version);
}
```

This drops the unnecessary promise wrap, because some resolve values are already a promise.